Query: Convert IS operator in predicate to comparison with discriminator

smitpatel opened the pull request:
Resolves #5007

smitpatel commented on src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs:
```diff
@@ -830,6 +835,68 @@ protected virtual void CheckClientEval([NotNull] object expression)
+        private static readonly MethodInfo _efPropertyMethod
+            = typeof(EF).GetTypeInfo().GetDeclaredMethod(nameof(EF.Property));
+
+        private static Expression CreatePropertyExpression(Expression target, IProperty property)
```
these methods are being used in 2 places (other one being `EntityEqualityRewritingExpressionVisitor`), any good place to refactor?

maumar replied:
Maybe EntityQueryModelVisitor? we already have a IsPropertyMethod on there

anpete replied:
Even introducing a new helper class would be preferable here.

maumar commented on src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs:
```diff
@@ -701,6 +703,9 @@ public override void VisitWhereClause(WhereClause whereClause, QueryModel queryM
 
             if (!requiresClientFilter)
             {
+                var discriminatorPredicate = new PedicateTypeIsExpressionTranslatingVisitor(this).Visit(whereClause.Predicate);
```
typo

maumar commented on test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs:
```diff
@@ -30,7 +30,18 @@ public override void Can_use_is_kiwi()
                 @"SELECT [a].[Species], [a].[CountryId], [a].[Discriminator], [a].[Name], [a].[EagleId], [a].[IsFlightless], [a].[Group], [a].[FoundOn]
 FROM [Animal] AS [a]
-WHERE [a].[Discriminator] IN (N'Kiwi', N'Eagle')",
+WHERE [a].[Discriminator] IN (N'Kiwi', N'Eagle') AND ([a].[Discriminator] = N'Kiwi')",
```
that's something we could probably optimize (in the future)

smitpatel replied:
filed #5230

maumar commented:
:shipit:

anpete commented on src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs:
```diff
@@ -830,6 +835,68 @@ protected virtual void CheckClientEval([NotNull] object expression)
             }
         }
 
+        private class PedicateTypeIsExpressionTranslatingVisitor : ExpressionVisitorBase
```
*Predicate*

anpete replied:
I don't think this should just be for predicates. Add some tests with *is* in the projection etc.

smitpatel commented:
Updated with feedback.
The translatingvisitor now works on whole query model so projection/predicate both are translated. (added test for both).
Also removed generation of `Where 1 = 1` in query. If predicate is true then we don't generate where clause.

anpete commented on src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs:
```diff
@@ -830,6 +835,58 @@ protected virtual void CheckClientEval([NotNull] object expression)
+        {
+            private readonly RelationalQueryModelVisitor _queryModelVisitor;
+
+            public TypeIsExpressionTranslatingVisitor(RelationalQueryModelVisitor relationalQueryModelVisitor)
```
Minor but consider reducing coupling here by only passing the required dependencies: Model and RelationalAnnotationProvider. Then you can revert the RelationalAnnotationProvider property change, too.

anpete commented on src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/EntityEqualityRewritingExpressionVisitor.cs:
```diff
@@ -81,27 +81,17 @@ var newRightExpression
                             .ToArray()));
         }
 
-        private static readonly MethodInfo _efPropertyMethod
```
Nice.

smitpatel commented:
updated.

anpete commented:
:shipit:
